beluga/include/beluga/actions/propagate.hpp

```diff
@@ -0,0 +1,113 @@
+   * \param policy The execution policy to use.
+   * \param range An existing range of particles to apply this action to.
+   * \param model A callable instance to compute the states from the previous states.
+   */
+  template <
+      class ExecutionPolicy,
+      class Range,
+      class Model,
+      std::enable_if_t<std::is_execution_policy_v<std::decay_t<ExecutionPolicy>>, int> = 0,
+      std::enable_if_t<ranges::range<Range>, int> = 0>
+  constexpr auto operator()(ExecutionPolicy&& policy, Range& range, Model model) const -> Range& {
+    static_assert(beluga::is_particle_range_v<Range>);
+    auto states = range | beluga::views::states | ranges::views::common;
+    std::transform(
+        policy,              // rvalue policies are not supported in some STL implementations
+        std::begin(states),  //
+        std::end(states),    //
+        std::begin(states),  //
+        std::move(model));
+    return range;
+  }
+
+  /// Overload that re-orders arguments from a view closure.
+  template <
+      class Range,
+      class Model,
+      class ExecutionPolicy,
+      std::enable_if_t<ranges::range<Range>, int> = 0,
+      std::enable_if_t<std::is_execution_policy_v<ExecutionPolicy>, int> = 0>
+  constexpr auto operator()(Range&& range, Model model, ExecutionPolicy policy) const -> Range& {
+    return (*this)(std::move(policy), std::forward<Range>(range), std::move(model));
+  }
+
+  /// Overload that returns a view closure to compose with other views.
+  template <
+      class ExecutionPolicy,  //
+      class Model,            //
+      std::enable_if_t<std::is_execution_policy_v<ExecutionPolicy>, int> = 0>
+  constexpr auto operator()(ExecutionPolicy policy, Model model) const {
+    return ranges::make_action_closure(ranges::bind_back(propagate_base_fn{}, std::move(model), std::move(policy)));
+  }
+};
+
+/// Implementation detail for a propagate range adaptor object with a default execution policy.
+struct propagate_fn : public propagate_base_fn {
+  using propagate_base_fn::operator();
+
+  /// Overload that defines a default execution policy.
+  template <
+      class Range,  //
+      class Model,  //
+      std::enable_if_t<ranges::range<Range>, int> = 0>
+  constexpr auto operator()(Range&& range, Model model) const -> Range& {
+    return (*this)(std::execution::seq, std::forward<Range>(range), std::move(model));
+  }
+
+  /// Overload that returns a view closure to compose with other views.
+  template <class Model>
+  constexpr auto operator()(Model model) const {
+    return ranges::make_action_closure(ranges::bind_back(propagate_fn{}, std::move(model)));
+  }
+};
+
+}  // namespace detail
+
+/// [Range adaptor object](https://en.cppreference.com/w/cpp/named_req/RangeAdaptorObject) that
+/// can update the state in a particle range using a motion model.
+/**
+ * This step generates a hyphotetical state based on the current particle state and controls,
+ * while leaving its importance weight unchanged.
```

Comment:
@nahueespinosa nit: perhaps consider:
```suggestion
 * This action updates particle states based on their current value and a state-transition function (or motion model).
 * Every other particle attribute (such as importance sampling weights) is left unchanged.
```